xprof/tally.cpp

```diff
@@ -32,14 +38,46 @@ void print_metadata(std::vector<std::string> metadata) {
     std::cout << value << std::endl;
 }
 
+int get_backend_id(std::string name){
+  for(int i = 0; i < BACKEND_MAX; ++i)
+    // backend_name is located in xprof_utils.cpp
+    if (std::string{backend_name[i]} == name) return i;
+  return -1;
+}
+
 void btx_initialize_usr_data(void *btx_handle, void **usr_data) {
   /* User allocates its own data structure */
-  *usr_data = new tally_dispatch_t;
+  auto *data = new tally_dispatch_t;
+  *usr_data = data;
+
+  /* Backend information must match enum backend_e in xprof_utils.hpp */
+  data->backend_level = {
+    2, // BACKEND_UNKNOWN
+    2, // BACKEND_ZE
+    2, // BACKEND_OPENCL
+    2, // BACKEND_CUDA
+    1, // BACKEND_OMP_TARGET_OPERATIONS
+    0, // BACKEND_OMP
+    2, // BACKEND_HIP
+  };
 }
 
 void btx_read_params(void *btx_handle, void *usr_data, btx_params_t *usr_params) {
   tally_dispatch_t *data = (tally_dispatch_t *)usr_data;
   data->params = usr_params;
+
+  // Consumes key:value pairs in the stringstream k1:v1,..,kn:vn
+  std::stringstream tokens{data->params->backend_level};
+  std::string tmp;
+  while (getline(tokens, tmp, ',')) {
```

Comment:
Please use `std::getline` no idea why it's in the general namespace. Who is doing `using std::string` or `using std::cstring`?

Reply:
I did a quick search in the header files, sstream export the symbol, but it is scoped in the std name space. I did this quick example and it semmed to not be THAPI issue.

```c++
#include <iostream>
#include <string>

int main()
{
    std::string str;
    std::cout << "Please enter your name: \n";
    getline(std::cin, str);
    std::cout << "Hello, " << str  << " welcome to GfG !\n";

    return 0;
}
```